Gems/PhysX/Code/Source/HeightfieldColliderComponent.cpp

```diff
@@ -62,22 +63,79 @@ namespace PhysX
     {
     }
 
-    void HeightfieldColliderComponent::Activate()
+
+    void HeightfieldColliderComponent::OnAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset)
     {
-        const AZ::EntityId entityId = GetEntityId();
+        if (asset == m_bakedHeightfieldAsset)
+        {
+            m_bakedHeightfieldAsset = asset;
 
-        AzPhysics::SceneHandle sceneHandle = AzPhysics::InvalidSceneHandle;
-        Physics::DefaultWorldBus::BroadcastResult(sceneHandle, &Physics::DefaultWorldRequests::GetDefaultSceneHandle);
+            Physics::HeightfieldShapeConfiguration& configuration = static_cast<Physics::HeightfieldShapeConfiguration&>(*m_shapeConfig);
+            configuration = Utils::CreateBaseHeightfieldShapeConfiguration(GetEntityId());
 
-        m_heightfieldCollider =
-            AZStd::make_unique<HeightfieldCollider>(GetEntityId(), GetEntity()->GetName(), sceneHandle, m_colliderConfig, m_shapeConfig);
+            // Update material selection from the mapping
+            Physics::ColliderConfiguration* colliderConfig = m_colliderConfig.get();
+            Utils::SetMaterialsFromHeightfieldProvider(GetEntityId(), colliderConfig->m_materialSlots);
 
-        ColliderComponentRequestBus::Handler::BusConnect(entityId);
-        Physics::CollisionFilteringRequestBus::Handler::BusConnect(entityId);
+            Pipeline::HeightFieldAsset* heightfieldAsset = m_bakedHeightfieldAsset.Get();
+
+            bool minMaxHeightsMatch = AZ::IsClose(configuration.GetMinHeightBounds(), heightfieldAsset->GetMinHeight()) &&
+                AZ::IsClose(configuration.GetMaxHeightBounds(), heightfieldAsset->GetMaxHeight());
+
+            if (!minMaxHeightsMatch)
+            {
+                AZ_Warning(
+                    "PhysX",
+                    false,
+                    "MinMax heights mismatch between baked heightfield and terrain. Entity [%s]. "
+                    "Terrain [%0.2f, %0.2f], Asset [%0.2f, %0.2f]",
+                    GetEntity()->GetName().c_str(),
+                    configuration.GetMinHeightBounds(),
+                    configuration.GetMaxHeightBounds(),
+                    heightfieldAsset->GetMinHeight(),
+                    heightfieldAsset->GetMaxHeight());
+            }
+
+            physx::PxHeightField* pxHeightfield = heightfieldAsset->GetHeightField();
+
+            // Since PxHeightfield will have shared ownership in both HeightfieldAsset and HeightfieldShapeConfiguration,
+            // we need to increment the reference counter here. Both of these places call release() in destructors,
+            // so we need to avoid double deletion this way.
+            pxHeightfield->acquireReference();
+            configuration.SetCachedNativeHeightfield(pxHeightfield);
+
+            InitHeightfieldCollider();
```

Comment:
In the case of a baked asset, won't creating a HeightfieldCollider cause it to still listen for data changes and try to overwrite the baked data at runtime? This seems like it will eliminate any benefit of the baked asset, because I think it will still get refreshed exactly as much as without a baked asset.

Reply:
One thing to test with these changes is making sure that changes to the transform works correctly with baked heightfields. I'd think they should still be able to move. Transform notifications would likely need to be handled differently between baked and non-baked heightfields. Non-baked ones have an indirection because a transform change actually changes the heightfield data itself. Baked ones wouldn't need that indirection, they would just move the position of the rigid body I think.

As a side note, I haven't dug deep into exactly just how much of the HeightfieldCollider is short-circuited with these changes, but if it's nearly the whole thing, maybe it would make sense to give it an interface class and have a couple of subclasses for baked vs non-baked.